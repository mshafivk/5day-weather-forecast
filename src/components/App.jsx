import React from 'react';
import styles from './App.less';

const App = () => (
  <div className={styles.container}>
    <h2>Weather Forcast</h2>
  </div>
);
export default App;
