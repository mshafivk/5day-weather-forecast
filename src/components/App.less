@import '../common/color.less';
@import '../common//typography.less';

.container {
  background-color: @bg-color-primary;
  color: @fg-color-primary;
  display: block;
  height: 300px;
  padding: 8px;
  h2 {
    .header();
  }
}
