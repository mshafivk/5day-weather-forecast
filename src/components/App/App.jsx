import React, { useEffect, useReducer } from 'react';
import {
  SET_CITY,
  SET_LOADING_STATE,
  SET_FORECAST,
} from '../../reducers/actions';
import { WeatherWidget } from '../WeatherWidget/WeatherWidget';
import { weatherReducer, initialState } from '../../reducers/weatherReducer';
import { fetchWeatherInfo } from '../../common/services';
import classes from './App.less';

const App = () => {
  const [state, dispatch] = useReducer(weatherReducer, initialState);

  const handleCityChange = (evt) =>
    dispatch({ type: SET_CITY, payload: { city: evt.target.value } });

  const loadInformation = async () => {
    dispatch({ type: SET_LOADING_STATE, payload: { loading: true } });
    const response = await fetchWeatherInfo(state.city);
    dispatch({
      type: SET_FORECAST,
      payload: { forecasts: response.list || [] },
    });
    dispatch({ type: SET_LOADING_STATE, payload: { loading: false } });
  };

  useEffect(loadInformation, []);

  return (
    <div className={classes.container}>
      <h2>5 Day Weather Forcast</h2>
      <div>
        City :{' '}
        <input type="text" value={state.city} onChange={handleCityChange} />
        <button onClick={loadInformation}>UPDATE WEATHER</button>
      </div>
      <div className={classes.main}>
        {state.loading ? (
          <div data-testid="loading-indicator">Loading...</div>
        ) : (
          <WeatherWidget forecasts={state.forecasts} />
        )}
      </div>
    </div>
  );
};
export default App;
