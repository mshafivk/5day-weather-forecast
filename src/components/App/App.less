@import '../../common/styles/color.less';
@import '../../common/styles/typography.less';

.container {
  background-color: @bg-color-primary;
  color: @fg-color-primary;
  display: block;
  min-height: 300px;
  padding: 8px;
  h2 {
    .header();
    text-align: center;
  }

  .main {
    margin-top: 8px;
  }
}
