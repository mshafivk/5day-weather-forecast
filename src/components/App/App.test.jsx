import React from 'react';
import App from './App';
import { shallow } from 'enzyme';

describe('App Component', () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({ list: [] }),
    }),
  );

  beforeEach(() => {
    fetch.mockClear();
  });
  it('should render properly', () => {
    const wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should show loading message and invoke fetch on button click', () => {
    const wrapper = shallow(<App />);
    wrapper.find('button').simulate('click');
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(wrapper.find('[data-testid="loading-indicator"]')).toHaveLength(1);
  });
  test('if city can be updated', () => {
    const wrapper = shallow(<App />);
    wrapper.find('input').simulate('change', { target: { value: 'dummy' } });
    expect(wrapper.find('input').props().value).toEqual('dummy');
  });
});
