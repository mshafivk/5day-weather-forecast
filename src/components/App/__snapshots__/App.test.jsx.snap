// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render properly 1`] = `
<div>
  <h2>
    5 Day Weather Forcast
  </h2>
  <div>
    City :
     
    <input
      onChange={[Function]}
      type="text"
      value="BANGALORE"
    />
    <button
      onClick={[Function]}
    >
      UPDATE WEATHER
    </button>
  </div>
  <div>
    <WeatherWidget
      forecasts={Array []}
    />
  </div>
</div>
`;
